File: web/src/services/api/workflow-registry.ts
import { apiGet, apiPatch, apiPost, apiPostEmpty, apiPut, type ApiParams } from "@/services/api/request";
import { useUserStore } from "@/stores/use-user-store";

import type { ArtifactRefInput } from "./invocations-contract";

export type WorkflowOwnerType = "system" | "project";
export type WorkflowVersionStatus = "draft" | "published" | "retired";
export type WorkflowExecutorType = "skill" | "agent";
export type WorkflowSkillBindingMode = "fixed" | "tag_route" | "manual_before_run";
export type WorkflowExecutionStatus = "preflight" | "awaiting_confirmation" | "running" | "needs_review" | "completed" | "blocked" | "partial" | "failed" | "cancelled";
export type WorkflowNodeExecutionStatus = "blocked" | "ready" | "queued" | "running" | "needs_review" | "approved" | "completed" | "skipped" | "failed" | "cancelled";

export type WorkflowDefinition = {
    id: string;
    name: string;
    summary: string;
    ownerType: WorkflowOwnerType;
    ownerUserId: string;
    ownerProjectId: string;
    enabled: boolean;
    recommendedVersionId: string;
    createdAt: string;
    updatedAt: string;
};

export type WorkflowVersion = {
    id: string;
    workflowId: string;
    version: string;
    status: WorkflowVersionStatus;
    contentHash: string;
    createdBy: string;
    publishedAt: string;
    createdAt: string;
    updatedAt: string;
};

export type WorkflowSkillBinding = {
    mode: WorkflowSkillBindingMode;
    skillId: string;
    skillVersionId: string;
    skillVersionConstraint: string;
    capability: string;
    expectedOutputArtifactType: string;
    projectTags: string[];
    candidateSkillIds: string[];
};

export type WorkflowAgentRef = { agentId: string; agentVersionId: string; agentVersionConstraint: string };
export type WorkflowNodeInputBinding = {
    bindingName: string;
    artifactType: string;
    source: "workflow_input" | "node_output";
    workflowInputName?: string;
    fromNodeKey?: string;
    fromOutputBinding?: string;
    required: boolean;
};
export type WorkflowCondition = { source: "workflow_input" | "node_output"; key: string; operator: "equals" | "not_equals" | "contains" | "exists"; value?: unknown };
export type WorkflowNodeSpec = {
    nodeKey: string;
    name: string;
    executorType: WorkflowExecutorType;
    agentRef?: WorkflowAgentRef;
    skillBinding?: WorkflowSkillBinding;
    inputBindings: WorkflowNodeInputBinding[];
    outputArtifactType: string;
    dependsOn: string[];
    condition?: WorkflowCondition;
    confirmationPolicy: { requireBeforeRun: boolean; requireReview: boolean };
    retryPolicy: { maxAttempts: number };
};
export type WorkflowPackage = { inputArtifactTypes: string[]; nodes: WorkflowNodeSpec[]; contentHash: string };
export type WorkflowVersionDetail = { workflow: WorkflowDefinition; version: WorkflowVersion; package: WorkflowPackage; tags: string[] };
export type WorkflowRegistryItem = { workflow: WorkflowDefinition; tags: string[]; versions: WorkflowVersion[]; recommendedPackage?: WorkflowPackage };
export type WorkflowCreateInput = { projectId: string; name: string; summary: string; tags: string[]; version: string; package: WorkflowPackage };
export type WorkflowDraftInput = Pick<WorkflowCreateInput, "version" | "package">;
export type WorkflowValidationResult = { contentHash: string; resolvedNodes: Array<{ nodeKey: string; executorType: WorkflowExecutorType; agentId?: string; agentVersionId?: string; skillId?: string; skillVersionId?: string; skillContentHash?: string }> };

export type WorkflowRouteCandidate = { skillId: string; skillVersionId: string; accepted: boolean; score: number; reasons: string[] };
export type WorkflowRouteTrace = { capability: string; candidates: WorkflowRouteCandidate[]; finalSkillVersionId: string; selectedModel: string };
export type WorkflowNodeRoutePreview = {
    nodeKey: string;
    name: string;
    executorType: WorkflowExecutorType;
    agentVersionId?: string;
    skillVersionId?: string;
    skillContentHash?: string;
    routeTrace: WorkflowRouteTrace;
    estimatedCredits: number;
    confirmationCodes: string[];
    blockCode?: string;
    blockMessage?: string;
};
export type WorkflowRoutePreview = { workflowVersionId: string; contentHash: string; executable: boolean; estimatedCredits: number; confirmationRequirements: string[]; nodes: WorkflowNodeRoutePreview[] };
export type WorkflowPreviewInput = { projectId: string; episodeId?: string; inputArtifactRefs: ArtifactRefInput[]; manualSelections: Record<string, string>; projectTags: string[]; parameters: Record<string, unknown> };

export type WorkflowExecutionPreflightInput = WorkflowPreviewInput & { workflowVersionId: string; idempotencyKey: string };
export type WorkflowExecutionConfirmationInput = { revision: number; fingerprint: string; requirementCodes: string[] };
export type WorkflowExecutionRun = {
    id: string;
    projectId: string;
    episodeId: string;
    workflowId: string;
    workflowVersionId: string;
    workflowContentHash: string;
    status: WorkflowExecutionStatus;
    revision: number;
    estimatedCredits: number;
    confirmationFingerprint: string;
    createdAt: string;
    updatedAt: string;
};
export type WorkflowExecutionRevision = {
    id: string;
    workflowExecutionId: string;
    revision: number;
    workflowVersionId: string;
    workflowContentHash: string;
    inputArtifactRefs: ArtifactRefInput[];
    manualSelections: Record<string, string>;
    parameters: Record<string, unknown> | null;
    estimatedCredits: number;
    confirmationFingerprint: string;
    createdAt: string;
};
export type WorkflowNodeExecution = {
    id: string;
    workflowExecutionId: string;
    revision: number;
    ordinal: number;
    nodeKey: string;
    executorType: WorkflowExecutorType;
    invocationId?: string;
    agentPlanId?: string;
    status: WorkflowNodeExecutionStatus;
    outputArtifactRefs: ArtifactRefInput[];
    errorCode?: string;
    errorMessage?: string;
    createdAt: string;
    updatedAt: string;
};
export type WorkflowExecutionConfirmation = { id: string; workflowExecutionId: string; revision: number; fingerprint: string; estimatedCredits: number; confirmedAt: string };
export type WorkflowExecutionResponse = {
    run: WorkflowExecutionRun;
    revision: WorkflowExecutionRevision;
    nodes: WorkflowNodeExecution[];
    preview: WorkflowRoutePreview;
    confirmationRequirements: string[];
    confirmation?: WorkflowExecutionConfirmation;
};

export type WorkflowApiGet = <T>(url: string, params?: ApiParams, token?: string) => Promise<T>;
export type WorkflowApiPost = <T>(url: string, body?: unknown, token?: string) => Promise<T>;
export type WorkflowApiPostEmpty = <T>(url: string, token?: string) => Promise<T>;
export type WorkflowApiPatch = <T>(url: string, body?: unknown, token?: string) => Promise<T>;
export type WorkflowApiPut = <T>(url: string, body?: unknown, token?: string) => Promise<T>;

export function createWorkflowRegistryClient(deps: { apiGet: WorkflowApiGet; apiPost: WorkflowApiPost; apiPostEmpty: WorkflowApiPostEmpty; apiPatch: WorkflowApiPatch; apiPut: WorkflowApiPut; token: () => string }) {
    const auth = () => deps.token();
    const id = encodeURIComponent;
    return {
        fetchWorkflows: (projectId: string) => deps.apiGet<WorkflowRegistryItem[]>("/api/v1/workflows", { projectId }, auth()),
        fetchWorkflow: (workflowId: string, projectId: string) => deps.apiGet<WorkflowRegistryItem>(`/api/v1/workflows/${id(workflowId)}`, { projectId }, auth()),
        createWorkflow: (input: WorkflowCreateInput) => deps.apiPost<WorkflowVersionDetail>("/api/v1/workflows", input, auth()),
        copyWorkflow: (workflowId: string, projectId: string, name: string) => deps.apiPost<WorkflowVersionDetail>(`/api/v1/workflows/${id(workflowId)}/copy`, { projectId, name }, auth()),
        createWorkflowVersion: (workflowId: string, input: WorkflowDraftInput) => deps.apiPost<WorkflowVersion>(`/api/v1/workflows/${id(workflowId)}/versions`, input, auth()),
        fetchWorkflowVersion: (versionId: string) => deps.apiGet<WorkflowVersionDetail>(`/api/v1/workflow-versions/${id(versionId)}`, undefined, auth()),
        updateWorkflowVersion: (versionId: string, input: WorkflowDraftInput) => deps.apiPatch<WorkflowVersion>(`/api/v1/workflow-versions/${id(versionId)}`, input, auth()),
        validateWorkflowVersion: (versionId: string) => deps.apiPostEmpty<WorkflowValidationResult>(`/api/v1/workflow-versions/${id(versionId)}/validate`, auth()),
        previewWorkflowVersion: (versionId: string, input: WorkflowPreviewInput) => deps.apiPost<WorkflowRoutePreview>(`/api/v1/workflow-versions/${id(versionId)}/preview`, input, auth()),
        publishWorkflowVersion: (versionId: string) => deps.apiPostEmpty<WorkflowVersionDetail>(`/api/v1/workflow-versions/${id(versionId)}/publish`, auth()),
        recommendWorkflowVersion: (workflowId: string, workflowVersionId: string) => deps.apiPut<WorkflowVersionDetail>(`/api/v1/workflows/${id(workflowId)}/recommended-version`, { workflowVersionId }, auth()),
        preflightWorkflowExecution: (input: WorkflowExecutionPreflightInput) => deps.apiPost<WorkflowExecutionResponse>("/api/v1/workflow-executions/preflight", input, auth()),
        fetchWorkflowExecution: (executionId: string) => deps.apiGet<WorkflowExecutionResponse>(`/api/v1/workflow-executions/${id(executionId)}`, undefined, auth()),
        confirmWorkflowExecution: (executionId: string, input: WorkflowExecutionConfirmationInput) => deps.apiPost<WorkflowExecutionResponse>(`/api/v1/workflow-executions/${id(executionId)}/confirm`, input, auth()),
        continueWorkflowExecution: (executionId: string) => deps.apiPostEmpty<WorkflowExecutionResponse>(`/api/v1/workflow-executions/${id(executionId)}/continue`, auth()),
        cancelWorkflowExecution: (executionId: string) => deps.apiPostEmpty<WorkflowExecutionResponse>(`/api/v1/workflow-executions/${id(executionId)}/cancel`, auth()),
    };
}

export const workflowRegistryClient = createWorkflowRegistryClient({ apiGet, apiPost, apiPostEmpty, apiPatch, apiPut, token: () => useUserStore.getState().token });
export const {
    fetchWorkflows, fetchWorkflow, createWorkflow, copyWorkflow, createWorkflowVersion, fetchWorkflowVersion, updateWorkflowVersion,
    validateWorkflowVersion, previewWorkflowVersion, publishWorkflowVersion, recommendWorkflowVersion,
    preflightWorkflowExecution, fetchWorkflowExecution, confirmWorkflowExecution, continueWorkflowExecution, cancelWorkflowExecution,
} = workflowRegistryClient;

File: model/workflow_registry.go
package model

type WorkflowOwnerType string
type WorkflowVersionStatus string
type WorkflowExecutionStatus string
type WorkflowNodeExecutionStatus string

const (
	WorkflowOwnerSystem  WorkflowOwnerType = "system"
	WorkflowOwnerProject WorkflowOwnerType = "project"

	WorkflowVersionDraft     WorkflowVersionStatus = "draft"
	WorkflowVersionPublished WorkflowVersionStatus = "published"
	WorkflowVersionRetired   WorkflowVersionStatus = "retired"

	WorkflowExecutionPreflight            WorkflowExecutionStatus = "preflight"
	WorkflowExecutionAwaitingConfirmation WorkflowExecutionStatus = "awaiting_confirmation"
	WorkflowExecutionRunning              WorkflowExecutionStatus = "running"
	WorkflowExecutionNeedsReview          WorkflowExecutionStatus = "needs_review"
	WorkflowExecutionCompleted            WorkflowExecutionStatus = "completed"
	WorkflowExecutionBlocked              WorkflowExecutionStatus = "blocked"
	WorkflowExecutionPartial              WorkflowExecutionStatus = "partial"
	WorkflowExecutionFailed               WorkflowExecutionStatus = "failed"
	WorkflowExecutionCancelled            WorkflowExecutionStatus = "cancelled"

	WorkflowNodeExecutionBlocked     WorkflowNodeExecutionStatus = "blocked"
	WorkflowNodeExecutionReady       WorkflowNodeExecutionStatus = "ready"
	WorkflowNodeExecutionQueued      WorkflowNodeExecutionStatus = "queued"
	WorkflowNodeExecutionRunning     WorkflowNodeExecutionStatus = "running"
	WorkflowNodeExecutionNeedsReview WorkflowNodeExecutionStatus = "needs_review"
	WorkflowNodeExecutionApproved    WorkflowNodeExecutionStatus = "approved"
	WorkflowNodeExecutionCompleted   WorkflowNodeExecutionStatus = "completed"
	WorkflowNodeExecutionSkipped     WorkflowNodeExecutionStatus = "skipped"
	WorkflowNodeExecutionFailed      WorkflowNodeExecutionStatus = "failed"
	WorkflowNodeExecutionCancelled   WorkflowNodeExecutionStatus = "cancelled"
)

type WorkflowDefinition struct {
	ID                   string            `json:"id" gorm:"primaryKey"`
	Name                 string            `json:"name" gorm:"index;uniqueIndex:idx_workflow_owner_name,priority:4"`
	Summary              string            `json:"summary" gorm:"type:text"`
	TagsJSON             string            `json:"-" gorm:"type:text"`
	OwnerType            WorkflowOwnerType `json:"ownerType" gorm:"index;uniqueIndex:idx_workflow_owner_name,priority:1"`
	OwnerUserID          string            `json:"ownerUserId" gorm:"index;uniqueIndex:idx_workflow_owner_name,priority:2"`
	OwnerProjectID       string            `json:"ownerProjectId" gorm:"index;uniqueIndex:idx_workflow_owner_name,priority:3"`
	Enabled              bool              `json:"enabled" gorm:"index"`
	RecommendedVersionID string            `json:"recommendedVersionId" gorm:"index"`
	CreatedAt            string            `json:"createdAt"`
	UpdatedAt            string            `json:"updatedAt"`
}

type WorkflowVersion struct {
	ID          string                `json:"id" gorm:"primaryKey"`
	WorkflowID  string                `json:"workflowId" gorm:"index;uniqueIndex:idx_workflow_version,priority:1"`
	Version     string                `json:"version" gorm:"uniqueIndex:idx_workflow_version,priority:2"`
	Status      WorkflowVersionStatus `json:"status" gorm:"index"`
	PackageJSON string                `json:"-" gorm:"type:text"`
	ContentHash string                `json:"contentHash" gorm:"index"`
	CreatedBy   string                `json:"createdBy" gorm:"index"`
	PublishedAt string                `json:"publishedAt"`
	CreatedAt   string                `json:"createdAt"`
	UpdatedAt   string                `json:"updatedAt"`
}

type WorkflowExecution struct {
	ID                      string                  `json:"id" gorm:"primaryKey"`
	UserID                  string                  `json:"userId" gorm:"index;uniqueIndex:idx_workflow_execution_idempotency,priority:1"`
	ProjectID               string                  `json:"projectId" gorm:"index"`
	EpisodeID               string                  `json:"episodeId" gorm:"index"`
	WorkflowID              string                  `json:"workflowId" gorm:"index"`
	WorkflowVersionID       string                  `json:"workflowVersionId" gorm:"index"`
	WorkflowContentHash     string                  `json:"workflowContentHash" gorm:"index"`
	Status                  WorkflowExecutionStatus `json:"status" gorm:"index"`
	Revision                int                     `json:"revision"`
	EstimatedCredits        int64                   `json:"estimatedCredits"`
	IdempotencyKey          *string                 `json:"-" gorm:"uniqueIndex:idx_workflow_execution_idempotency,priority:2"`
	RequestHash             string                  `json:"-" gorm:"index"`
	ConfirmationFingerprint string                  `json:"confirmationFingerprint" gorm:"index"`
	CreatedAt               string                  `json:"createdAt"`
	UpdatedAt               string                  `json:"updatedAt"`
}

type WorkflowExecutionRevision struct {
	ID                           string `json:"id" gorm:"primaryKey"`
	UserID                       string `json:"userId" gorm:"index"`
	WorkflowExecutionID          string `json:"workflowExecutionId" gorm:"index;uniqueIndex:idx_workflow_execution_revision,priority:1"`
	Revision                     int    `json:"revision" gorm:"uniqueIndex:idx_workflow_execution_revision,priority:2"`
	WorkflowVersionID            string `json:"workflowVersionId" gorm:"index"`
	WorkflowContentHash          string `json:"workflowContentHash" gorm:"index"`
	RoutePreviewJSON             string `json:"-" gorm:"type:text"`
	InputArtifactRefsJSON        string `json:"-" gorm:"type:text"`
	ManualSelectionsJSON         string `json:"-" gorm:"type:text"`
	EstimatedCredits             int64  `json:"estimatedCredits"`
	ConfirmationRequirementsJSON string `json:"-" gorm:"type:text"`
	ConfirmationFingerprint      string `json:"confirmationFingerprint" gorm:"index"`
	CreatedAt                    string `json:"createdAt"`
}

type WorkflowNodeExecution struct {
	ID                     string                      `json:"id" gorm:"primaryKey"`
	UserID                 string                      `json:"userId" gorm:"index"`
	WorkflowExecutionID    string                      `json:"workflowExecutionId" gorm:"index;uniqueIndex:idx_workflow_node_execution,priority:1"`
	Revision               int                         `json:"revision" gorm:"uniqueIndex:idx_workflow_node_execution,priority:2"`
	Ordinal                int                         `json:"ordinal"`
	NodeKey                string                      `json:"nodeKey" gorm:"index;uniqueIndex:idx_workflow_node_execution,priority:3"`
	ExecutorType           string                      `json:"executorType" gorm:"index"`
	InvocationID           string                      `json:"invocationId" gorm:"index"`
	AgentPlanID            string                      `json:"agentPlanId" gorm:"index"`
	Status                 WorkflowNodeExecutionStatus `json:"status" gorm:"index"`
	OutputArtifactRefsJSON string                      `json:"-" gorm:"type:text"`
	ErrorCode              string                      `json:"errorCode" gorm:"index"`
	ErrorMessage           string                      `json:"errorMessage" gorm:"type:text"`
	CreatedAt              string                      `json:"createdAt"`
	UpdatedAt              string                      `json:"updatedAt"`
}

type WorkflowExecutionConfirmation struct {
	ID                   string `json:"id" gorm:"primaryKey"`
	UserID               string `json:"userId" gorm:"index"`
	WorkflowExecutionID  string `json:"workflowExecutionId" gorm:"index;uniqueIndex:idx_workflow_execution_confirmation,priority:1"`
	Revision             int    `json:"revision" gorm:"uniqueIndex:idx_workflow_execution_confirmation,priority:2"`
	Fingerprint          string `json:"fingerprint" gorm:"index"`
	EstimatedCredits     int64  `json:"estimatedCredits"`
	RequirementCodesJSON string `json:"-" gorm:"type:text"`
	ConfirmedAt          string `json:"confirmedAt"`
}
